chasm/registrable_component.go

```diff
@@ -0,0 +1,82 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package chasm
+
+import (
+	"reflect"
+)
+
+type (
+	RegistrableComponent struct {
+		name   string
+		goType reflect.Type
+
+		ephemeral     bool
+		singleCluster bool
+		shardingFn    func(EntityKey) string
+	}
+
+	RegistrableComponentOption func(*RegistrableComponent)
+)
+
+func NewRegistrableComponent[C Component](
+	name string,
+	opts ...RegistrableComponentOption,
+) *RegistrableComponent {
+	rc := &RegistrableComponent{
+		name:       name,
+		goType:     reflect.TypeFor[C](),
+		shardingFn: func(_ EntityKey) string { return "" },
```

Comment:
let's use a default shardingFn here. key.NamespaceID_key.BusinessID, same as what we have for workflows today.